packages/gatsby-cli/src/create-cli.js

```diff
@@ -103,10 +103,22 @@ function buildLocalCommands(cli, isLocalSite) {
           alias: `https`,
           type: `boolean`,
           describe: `Use HTTPS. See https://www.gatsbyjs.org/docs/local-https/ for an initial setup guide`,
+        })
+        .option(`build-dir`, {
+          alias: `buildDirectory`,
+          type: `string`,
+          default: `public`,
+          describe: `Set build directory. Defaults to public`,
         }),
     handler: handlerP(
       getCommandHandler(`develop`, (args, cmd) => {
         process.env.NODE_ENV = process.env.NODE_ENV || `development`
+
+        process.env.GATSBY_BUILD_DIR = path.resolve(args.buildDirectory) || `public`
```

Comment:
Don't you want to respect what someone might have set in `GATSBY_BUILD_DIR`?

Seems like you could do this by making the default for the option be `process.env.GATSBY_BUILD_DIR || 'public'`.

Then `buildDirectory` would resolve to `'custom'` here:
```
GATSBY_BUILD_DIR=custom gatsby develop
```

And `'overridden'` here:
```
GATSBY_BUILD_DIR=custom gatsby develop --build-dir=overridden
```

And `'public'` here:
```
gatsby develop
```

Reply:
@tschaub Hmm maybe I didn't get what you want to say but if someone set his env to 
```
GATSBY_BUILD_DIR = my_folder_custom
```
and start gatsby with 
```
gatsby build
```
or 
```
gatsby develop
```
the folder that would be considered would be **my_custom_folder** because the only logic that could override `GATSBY_BUILD_DIR` would be into the `--build-dir=` arg